cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -250,14 +278,18 @@ public boolean apply(RunRecordMeta record) {
   private RuntimeInfo createRuntimeInfo(Id.Program programId, TwillController controller, RunId runId) {
     try {
       ProgramDescriptor programDescriptor = store.loadProgram(programId);
-      ProgramController programController = createController(programDescriptor, controller, runId);
-      return programController == null ? null : new SimpleRuntimeInfo(programController, programId.toEntityId(),
-                                                                      controller.getRunId());
+      ImpersonatedTwillController twillController = new ImpersonatedTwillController(impersonator,
+                                                                                    programId.toEntityId(), controller);
+      ProgramController programController = createController(programDescriptor, twillController, runId);
+      return programController == null ? null : new SimpleRuntimeInfo(programController,
+                                                                      programId.toEntityId(),
+                                                                      twillController.getRunId());
     } catch (Exception e) {
       return null;
     }
   }
 
+  @Nullable
```

Comment:
According to the logic, it's possible. And the caller also checks for it, hence I add the annotation to be clear.